ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ChainSync/Client.hs

```diff
@@ -376,6 +390,7 @@ bracketChainSyncClient
         atomically $ do
           modifyTVar varHandles $ Map.insert peer ChainSyncClientHandle {
               cschGDDKill = throwTo tid DensityTooLow
+            , cschGsmCallback = const (pure ()) -- will be updated later
```

Comment:
Hmm, couldn't we directly initialize it here?

Reply:
The callback works on the LoP bucket which is not available at that point of the execution; but I agree that it is not super clean.

Reply:
Eg like this:
```diff
diff --git a/ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ChainSync/Client.hs b/ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ChainSync/Client.hs
index 8a0c1d57c0..63f334a4f0 100644
--- a/ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ChainSync/Client.hs
+++ b/ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ChainSync/Client.hs
@@ -352,19 +352,13 @@ bracketChainSyncClient
     csBucketConfig
     body
   =
-    bracket acquireHandle releaseHandle
+    LeakyBucket.execAgainstBucket'
+  $ \lopBucket ->
+    bracket (acquireHandle lopBucket) releaseHandle
   $ \varState ->
         withWatcher
             "ChainSync.Client.rejectInvalidBlocks"
-            (invalidBlockWatcher varState)
-      $ LeakyBucket.execAgainstBucket'
-      $ \lopBucket -> do
-            -- NOTE: If `updateLopBucketConfig` was in STM, then all the
-            -- following could be in the same STM transaction which would make
-            -- it significantly more robust.
-            initialGsmState <- atomically getGsmState
-            updateLopBucketConfig lopBucket initialGsmState
-            atomically $ modifyTVar varHandles $ Map.update (\csch -> Just $! csch{cschGsmCallback = updateLopBucketConfig lopBucket}) peer
+            (invalidBlockWatcher varState) $ do
             body ChainSyncStateView {
               csvSetCandidate =
               modifyTVar varState . \ c s -> s {csCandidate = c}
@@ -381,17 +375,22 @@ bracketChainSyncClient
               }
             }
   where
-    acquireHandle = do
+    acquireHandle lopBucket = do
         cschState <- newTVarIO $ ChainSyncState {
             csCandidate = AF.Empty AF.AnchorGenesis
           , csLatestSlot = Nothing
           , csIdling = False
           }
         tid <- myThreadId
+        -- NOTE: If `updateLopBucketConfig` was in STM, then all the
+        -- following could be in the same STM transaction which would make
+        -- it significantly more robust.
+        initialGsmState <- atomically getGsmState
+        updateLopBucketConfig lopBucket initialGsmState
         atomically $ do
           modifyTVar varHandles $ Map.insert peer ChainSyncClientHandle {
               cschGDDKill = throwTo tid DensityTooLow
-            , cschGsmCallback = const (pure ()) -- will be updated later
+            , cschGsmCallback = updateLopBucketConfig lopBucket
             , cschState
             }
         pure cschState
```

Reply:
Yes, I was wondering about that one. I was just not sure what that would mean for exceptions thrown before `bracket` has fully “installed” itself. But probably that just will not happen because the bucket will not run out that fast and there shouldn't be other exceptions that are not bugs.

Reply:
The bucket won't actually leak until we potentially change the config in the `bracket` allocate part, as you use `execAgainstBucket'` which uses `dummyConfig`, right?

Reply:
That's a fair point; now that we're using `execAgainstBucket'` it must be safe (and it probably was almost always safe before anyways).

Reply:
Done in 18140d8b298d6219be52512ad1efe3752d9dc102.